zerver/lib/validator.py

```diff
@@ -190,63 +202,74 @@ def check_list(sub_validator: Optional[Validator], length: Optional[int]=None) -
         type_structure = None  # type: ignore[assignment] # monkey-patching
 
     @set_type_structure(type_structure)
-    def f(var_name: str, val: object) -> Optional[str]:
+    def f(var_name: str, val: object) -> List[ResultT]:
         if not isinstance(val, list):
-            return _('{var_name} is not a list').format(var_name=var_name)
+            raise ValidationError(_('{var_name} is not a list').format(var_name=var_name))
 
         if length is not None and length != len(val):
-            return (_('{container} should have exactly {length} items').format(
+            raise ValidationError(_('{container} should have exactly {length} items').format(
                 container=var_name, length=length,
             ))
 
         if sub_validator:
             for i, item in enumerate(val):
                 vname = f'{var_name}[{i}]'
-                error = sub_validator(vname, item)
-                if error:
-                    return error
+                valid_item = sub_validator(vname, item)
+                assert item is valid_item
```

Comment:
Is this assert for mypy, or to avoid some type of bug?  Might be worth a comment, here and in `check_dict`

Reply:
This is to justify (to human readers) the correctness of the unchecked cast below. Commented.